### PR TITLE
Cloud functions

### PR DESCRIPTION
**Pull request by pauljohanneskraft**
# Cloud Functions

## :recycle: Current situation & Problem
Cloud function code has previously been located in the [ENGAGE-HF-iOS](https://github.com/StanfordBDHG/ENGAGE-HF-iOS) repository and will now move here. Further, the code should be optimized in terms of using TypeScript rather than JS, etc.

## :gear: Release Notes 
Add Cloud Functions

## :books: Documentation
TBD

## :white_check_mark: Testing
TBD

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by PSchmiedmayer — CHANGES_REQUESTED** [review]
@pauljohanneskraft Thank you for working on this today!

It would be great if we can pull over the deployment automation as well + add the NextJSTemplate setup for TS + Listing as we move over the elements to this repo.

Apart from this, it might make sense to already adjust the security rules according to the new data structure setup to ensure that we have those in place as we expand and fill the system with more functionality?

File: functions/src/index.ts
Comment (PSchmiedmayer) [inline functions/src/index.ts:1]: We will need to expand the README data structure with the expected structure & elements as processed by this cloud function; e.g., `/invitationCodes` and the `dateOfEnrollment` for a user.

**Review by PSchmiedmayer — CHANGES_REQUESTED** [review]

File: functions/src/index.ts
Comment (pauljohanneskraft) [reply functions/src/index.ts:1]: Created this issue: https://github.com/StanfordBDHG/ENGAGE-HF-Firebase/issues/6

**Comment by pauljohanneskraft** [comment]
> It would be great if we can pull over the deployment automation as well + add the NextJSTemplate setup for TS + Listing as we move over the elements to this repo.

I already moved most of the relevant NextJSTemplate setup for TS into this PR - it would be nice to get a review / thoughts from @arkadiuszbachorski on this.

I'm trying to move the deployment automation in this PR as well, but depending on how long it is going to take, I might extract it into its own issue later on.

> Apart from this, it might make sense to already adjust the security rules according to the new data structure setup to ensure that we have those in place as we expand and fill the system with more functionality?

Created #7

File: .github/workflows/lintandbuild.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+on: [pull_request]
+
+jobs:
+  deploy:
+    name: Lint and Build Functions
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
```
Comment (PSchmiedmayer) [inline .github/workflows/lintandbuild.yml:19]: Would be good to use v4 and generally check that we use the latest version of the actions.

File: firestore.rules
Comment (PSchmiedmayer) [inline firestore.rules:1]: Would be good to create an issue to adapt these security rules as we add more data and seeding logic.

File: functions/src/index.ts
Comment (PSchmiedmayer) [inline functions/src/index.ts:1]: Thanks for the improvements here!

**Review by PSchmiedmayer — APPROVED** [review]
Thank you @pauljohanneskraft; I only had a few smaller comments. Feel free to merge the PR once they are resolved; thanks for the improvements and moving this here! 🚀

File: firestore.rules
Comment (pauljohanneskraft) [reply firestore.rules:1]: Already exists here: #7

File: .github/workflows/lintandbuild.yml (hunk 1 above)
Comment (pauljohanneskraft) [reply .github/workflows/lintandbuild.yml]: updated both to latest major version (actions/checkout and actions/setup-node)

**Pull request merged**